File: 42.cpp
struct Solution {
    int trap(vector<int>& height) {
        int res= 0, l = 0, r = height.size() - 1, maxLeft = 0, maxRight = 0;
        while (l < r){
            maxLeft = max(maxLeft, height[l]);
            maxRight = max(maxRight, height[r]);
            if (maxLeft < maxRight){
                res += maxLeft - height[l];
                ++l;
            } else {
                res += maxRight - height[r];
                --r;
            }
        }
        return res;
    }
};

File: 42.py
class Solution:
    def trap(self, height: List[int]) -> int:
        l, r = 0, len(height) - 1
        maxHeightOnLeft, maxHeightOnRight = 0, 0
        res = 0
        while l < r:
            if height[l] < height[r]:
                res += max(0, maxHeightOnLeft - height[l])
                maxHeightOnLeft = max(maxHeightOnLeft, height[l])
                l += 1
            else:
                res += max(0, maxHeightOnRight - height[r])
                maxHeightOnRight = max(maxHeightOnRight, height[r])
                r -= 1
        return res
